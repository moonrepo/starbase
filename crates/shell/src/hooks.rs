pub enum Statement<'data> {
    PrependPath {
        paths: &'data [String],
        key: Option<&'data str>,
        orig_key: Option<&'data str>,
    },
    SetEnv {
        key: &'data str,
        value: &'data str,
    },
    UnsetEnv {
        key: &'data str,
    },
}

pub enum Hook {
    OnChangeDir { command: String, prefix: String },
}

impl Hook {
    pub fn get_info(&self) -> &str {
        match self {
            Hook::OnChangeDir { .. } => "on change directory",
        }
    }
}
