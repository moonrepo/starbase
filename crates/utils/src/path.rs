use std::char::REPLACEMENT_CHARACTER;
use std::ffi::OsStr;
use std::hash::{DefaultHasher, Hasher};
use std::path::{Component, Path, PathBuf};

/// Normalize separators in a path string to their OS specific separators.
/// On Unix this will be `/`, and on Windows `\`.
#[inline]
pub fn normalize_separators<T: AsRef<str>>(path: T) -> String {
    #[cfg(unix)]
    {
        path.as_ref().replace('\\', "/")
    }

    #[cfg(windows)]
    {
        path.as_ref().replace('/', "\\")
    }
}

/// Standardize separators in a path string to `/` for portability,
#[inline]
pub fn standardize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}

/// Format the provided name for use as an executable file.
/// On Windows this will append `.exe`, on Unix used as-is.
#[inline]
pub fn exe_name<T: AsRef<str>>(name: T) -> String {
    let name = name.as_ref();

    #[cfg(unix)]
    {
        name.into()
    }

    #[cfg(windows)]
    {
        if name.ends_with(".exe") {
            name.into()
        } else {
            format!("{name}.exe")
        }
    }
}

/// Encode a value by removing invalid characters for use within a path component.
pub fn encode_component<T: AsRef<OsStr>>(value: T) -> String {
    let mut output = String::new();

    for ch in value.as_ref().to_string_lossy().chars() {
        match ch {
            '@' | '*' | REPLACEMENT_CHARACTER => {
                // Skip these
            }
            '/' | ':' => {
                output.push('-');
            }
            _ => {
                output.push(ch);
            }
        }
    }

    output.trim_matches(['-', '.']).to_owned()
}

/// Hash a value that may contain special characters into a valid path component.
pub fn hash_component<T: AsRef<OsStr>>(value: T) -> String {
    let mut hasher = DefaultHasher::default();
    hasher.write(value.as_ref().as_encoded_bytes());

    format!("{}", hasher.finish())
}

/// Clean a path by removing and flattening unnecessary path components.
pub fn clean<T: AsRef<Path>>(path: T) -> PathBuf {
    // Based on https://gitlab.com/foo-jin/clean-path
    let mut components = path.as_ref().components().peekable();

    let mut cleaned = if let Some(c @ Component::Prefix(..)) = components.peek().cloned() {
        components.next();

        PathBuf::from(c.as_os_str())
    } else {
        PathBuf::new()
    };

    let mut leading_parent_dots = 0;
    let mut component_count = 0;

    for component in components {
        match component {
            Component::Prefix(_) | Component::CurDir => {}
            Component::RootDir => {
                cleaned.push(component.as_os_str());
                component_count += 1;
            }
            Component::ParentDir => {
                if component_count == 1 && cleaned.is_absolute() {
                    // Nothing
                } else if component_count == leading_parent_dots {
                    cleaned.push("..");
                    leading_parent_dots += 1;
                    component_count += 1;
                } else {
                    cleaned.pop();
                    component_count -= 1;
                }
            }
            Component::Normal(c) => {
                cleaned.push(c);
                component_count += 1;
            }
        }
    }

    if component_count == 0 {
        cleaned.push(".");
    }

    cleaned
}

/// Return true if both provided paths are equal. Both paths will be cleaned
/// before comparison for accurate matching.
pub fn are_equal<L: AsRef<Path>, R: AsRef<Path>>(left: L, right: R) -> bool {
    clean(left) == clean(right)
}

/// Extend the native [`Path`] and [`PathBuf`] with additional functionality.
pub trait PathExt {
    /// Clean a path by removing and flattening unnecessary path components.
    fn clean(&self) -> PathBuf;

    /// Return true if the current path matches the provided path.
    /// Both paths will be cleaned before comparison for accurate matching.
    fn matches(&self, other: &Path) -> bool;
}

impl PathExt for Path {
    fn clean(&self) -> PathBuf {
        clean(self)
    }

    fn matches(&self, other: &Path) -> bool {
        are_equal(self, other)
    }
}

impl PathExt for PathBuf {
    fn clean(&self) -> PathBuf {
        clean(self)
    }

    fn matches(&self, other: &Path) -> bool {
        are_equal(self, other)
    }
}
