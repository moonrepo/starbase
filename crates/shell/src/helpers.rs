use std::env;
use std::ffi::OsString;
use std::path::{Path, PathBuf};

#[cfg(unix)]
pub static PATH_DELIMITER: &str = ":";

#[cfg(windows)]
pub static PATH_DELIMITER: &str = ";";

pub fn is_absolute_dir(value: OsString) -> Option<PathBuf> {
    let dir = PathBuf::from(&value);

    if !value.is_empty() && dir.is_absolute() {
        Some(dir)
    } else {
        None
    }
}

pub fn get_config_dir(home_dir: &Path) -> PathBuf {
    env::var_os("XDG_CONFIG_HOME")
        .and_then(is_absolute_dir)
        .unwrap_or_else(|| home_dir.join(".config"))
}

pub fn get_env_var_regex() -> regex::Regex {
    regex::Regex::new(r"\$(?<name>[A-Z0-9_]+)").unwrap()
}

pub fn normalize_newlines(content: impl AsRef<str>) -> String {
    let content = content.as_ref().trim();

    #[cfg(windows)]
    {
        content.replace('\r', "").replace('\n', "\r\n")
    }

    #[cfg(unix)]
    {
        content.replace('\r', "")
    }
}
