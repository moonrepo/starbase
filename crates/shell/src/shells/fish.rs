use super::Shell;
use crate::helpers::{get_config_dir, normalize_newlines};
use crate::hooks::*;
use std::collections::HashSet;
use std::fmt;
use std::path::{Path, PathBuf};

#[derive(Clone, Copy, Debug)]
pub struct Fish;

impl Fish {
    #[allow(clippy::new_without_default)]
    pub fn new() -> Self {
        Self
    }
}

// https://fishshell.com/docs/current/language.html#configuration
impl Shell for Fish {
    fn format(&self, statement: Statement<'_>) -> String {
        match statement {
            Statement::PrependPath {
                paths,
                key,
                orig_key,
            } => {
                let key = key.unwrap_or("PATH");
                let orig_key = orig_key.unwrap_or(key);

                format!(
                    r#"set -gx {key} {} ${orig_key};"#,
                    paths
                        .iter()
                        .map(|p| self.quote(p))
                        .collect::<Vec<_>>()
                        .join(" ")
                )
            }
            Statement::SetEnv { key, value } => {
                format!("set -gx {} {};", key, self.quote(value))
            }
            Statement::UnsetEnv { key } => {
                format!("set -ge {key};")
            }
        }
    }

    fn format_hook(&self, hook: Hook) -> Result<String, crate::ShellError> {
        Ok(normalize_newlines(match hook {
            Hook::OnChangeDir { command, prefix } => {
                format!(
                    r#"
set -gx __ORIG_PATH $PATH

function __{prefix}_hook --on-variable PWD;
  {command} | source
end;
"#
                )
            }
        }))
    }

    fn get_config_path(&self, home_dir: &Path) -> PathBuf {
        get_config_dir(home_dir).join("fish").join("config.fish")
    }

    fn get_env_path(&self, home_dir: &Path) -> PathBuf {
        self.get_config_path(home_dir)
    }

    fn get_profile_paths(&self, home_dir: &Path) -> Vec<PathBuf> {
        HashSet::<PathBuf>::from_iter([
            get_config_dir(home_dir).join("fish").join("config.fish"),
            home_dir.join(".config").join("fish").join("config.fish"),
        ])
        .into_iter()
        .collect()
    }

    /// Quotes a string according to Fish shell quoting rules.
    /// @see <https://fishshell.com/docs/current/language.html#quotes>
    fn quote(&self, value: &str) -> String {
        if value.is_empty() {
            return "''".to_string();
        }

        // Characters that need to be escaped in double quotes
        let escape_chars: &[(char, &str)] = &[
            ('\\', "\\\\"),
            ('\n', "\\n"),
            ('\t', "\\t"),
            ('\x07', "\\a"),
            ('\x08', "\\b"),
            ('\x1b', "\\e"),
            ('\x0c', "\\f"),
            ('\x0a', "\\n"),
            ('\x0d', "\\r"),
            ('\x0b', "\\v"),
            ('*', "\\*"),
            ('?', "\\?"),
            ('~', "\\~"),
            ('#', "\\#"),
            ('(', "\\("),
            (')', "\\)"),
            ('{', "\\{"),
            ('}', "\\}"),
            ('[', "\\["),
            (']', "\\]"),
            ('<', "\\<"),
            ('>', "\\>"),
            ('^', "\\^"),
            ('&', "\\&"),
            ('|', "\\|"),
            (';', "\\;"),
            ('"', "\\\""),
            // ('$', "\\$"),
        ];

        let mut quoted = value.to_string();
        for &(char, escape) in escape_chars.iter() {
            quoted = quoted.replace(char, escape);
        }

        if quoted.contains(' ') {
            format!("'{}'", quoted.replace('\'', "''"))
        } else {
            format!(r#""{}""#, quoted)
        }
    }
}
impl fmt::Display for Fish {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "fish")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use starbase_sandbox::assert_snapshot;

    #[test]
    fn formats_env_var() {
        assert_eq!(
            Fish.format_env_set("PROTO_HOME", "$HOME/.proto"),
            r#"set -gx PROTO_HOME "$HOME/.proto";"#
        );
    }

    #[test]
    fn formats_path() {
        assert_eq!(
            Fish.format_path_set(&["$PROTO_HOME/shims".into(), "$PROTO_HOME/bin".into()]),
            r#"set -gx PATH "$PROTO_HOME/shims" "$PROTO_HOME/bin" $PATH;"#
        );
    }

    #[test]
    fn formats_cd_hook() {
        let hook = Hook::OnChangeDir {
            command: "starbase hook fish".into(),
            prefix: "starbase".into(),
        };

        assert_snapshot!(Fish.format_hook(hook).unwrap());
    }

    #[test]
    fn test_fish_quoting() {
        assert_eq!(Fish.quote("\n"), r#""\n""#);
        assert_eq!(Fish.quote("\t"), r#""\t""#);
        assert_eq!(Fish.quote("\x07"), r#""\a""#);
        assert_eq!(Fish.quote("\x08"), r#""\b""#);
        assert_eq!(Fish.quote("\x1b"), r#""\e""#);
        assert_eq!(Fish.quote("\x0c"), r#""\f""#);
        assert_eq!(Fish.quote("\r"), r#""\r""#);
        assert_eq!(Fish.quote("\x0a"), r#""\n""#);
        assert_eq!(Fish.quote("\x0b"), r#""\v""#);
        assert_eq!(Fish.quote("*"), r#""\*""#);
        assert_eq!(Fish.quote("?"), r#""\?""#);
        assert_eq!(Fish.quote("~"), r#""\~""#);
        assert_eq!(Fish.quote("#"), r#""\#""#);
        assert_eq!(Fish.quote("("), r#""\(""#);
        assert_eq!(Fish.quote(")"), r#""\)""#);
        assert_eq!(Fish.quote("{"), r#""\{""#);
        assert_eq!(Fish.quote("}"), r#""\}""#);
        assert_eq!(Fish.quote("["), r#""\[""#);
        assert_eq!(Fish.quote("]"), r#""\]""#);
        assert_eq!(Fish.quote("<"), r#""\<""#);
        assert_eq!(Fish.quote(">"), r#""\>""#);
        assert_eq!(Fish.quote("^"), r#""\^""#);
        assert_eq!(Fish.quote("&"), r#""\&""#);
        assert_eq!(Fish.quote("|"), r#""\|""#);
        assert_eq!(Fish.quote(";"), r#""\;""#);
        assert_eq!(Fish.quote("\""), r#""\"""#);
        // assert_eq!(Fish.quote("$"), r#""\$""#);
        // assert_eq!(Fish.quote("$variable"), r#""\$variable""#);
        assert_eq!(Fish.quote("value with spaces"), "'value with spaces'");
    }
}
