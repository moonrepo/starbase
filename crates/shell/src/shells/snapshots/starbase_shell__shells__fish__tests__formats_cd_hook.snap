---
source: crates/shell/src/shells/fish.rs
expression: Fish.format_hook(hook).unwrap()
---
set -gx __ORIG_PATH $PATH

function __starbase_hook --on-variable PWD;
  starbase hook fish | source
end;
